### PR TITLE
write metric in tmp worker thread

### PR DESCRIPTION
**Pull request by lixinqi**
使用oneflow.async.thread将write_metrics放置在临时线程里，这样能加速eager的性能。加速效果大概是
| 测试模型   | 计算设备 | 单卡batch_size | 分布式情况 | oneflow性能     | pytorch性能    |
| ---------- | -------- | -------------- | ---------- | --------------- | -------------- |
| bert-large | v100     | 1              | 单卡       | 7.24 samples/s  | 6.8 samples/s  |
| bert-large | v100     | 2              | 单卡       | 8.62 samples/s  | 8.3 samples/s  |
| bert-large | v100     | 4              | 单卡       | 9.5 samples/s   | 9.2 samples/s  |
| bert-large | v100     | 1              | 两卡       | 11.54 samples/s | 9.5 samples/s  |
| bert-large | v100     | 2              | 两卡       | 15.19 samples/s | 13 samples/s   |
| bert-large | v100     | 4              | 两卡       | 17.77 samples/s | 16.0 samples/s |

若不做这个优化，每一项都会比pytorch要慢。

### CODE REVIEW

**Review by strint — APPROVED** [review]

**Comment by CPFLAME** [comment]
可以在四卡的机器上 本地跑一下  

`bash dev/model_test.sh` 

测试一下是否可以跑通模型的测试

**Comment by lixinqi** [comment]
> 可以在四卡的机器上 本地跑一下
> 
> `bash dev/model_test.sh`
> 
> 测试一下是否可以跑通模型的测试

@ouyangyu

File: libai/engine/trainer.py [hunk 1]
```diff
@@ -273,16 +274,35 @@ def run_step(self, get_batch: Callable, input_placement_device: str = "cuda"):
...
+        with self.thread(tmp_worker_thread):
+            self.write_metrics(tmp_loss_dict, data_time)
 
         if (self.iter + 1) % self.grad_acc_steps == 0:
             self.optimizer.clip_grad()
             self.optimizer.step()
             self.optimizer.zero_grad()
 
+    def tmp_worker_thread(self):
+        if not hasattr(flow, 'asyncs'):
```
Comment (hjchen2) [inline libai/engine/trainer.py:292]: 用try except是不是会更好看点

Comment (lixinqi) [reply libai/engine/trainer.py:292]: 有道理
